### PR TITLE
[Jenkins 31202]

### PR DESCRIPTION
**Pull request by benjaminfuchs**
https://issues.jenkins-ci.org/browse/JENKINS-31202

Required changes to show trend graphs in workflow job view.

This is related to the PR for analysis-core. I updated the WarningsPublisherTest and created a new one to test the new feature. To run the test I also had to update analysis-pom: https://github.com/benjaminfuchs/analysis-pom-plugin/tree/JENKINS-31202 (maybe there is another way to get the tests running.)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -53,16 +53,60 @@
       <optional>true</optional>
     </dependency>
     <dependency>
-      <groupId>commons-io</groupId>
-      <artifactId>commons-io</artifactId>
-      <version>2.2</version>
-    </dependency>
-    <dependency>
```
Comment (benjaminfuchs) [inline pom.xml]: This led to an problem with the test runs.

File: pom.xml [hunk 2]
```diff
@@ -53,16 +53,60 @@
...
+        </exclusion>
+      </exclusions>
+    </dependency>
+    <dependency>
+      <groupId>org.jenkins-ci.plugins.workflow</groupId>
+      <artifactId>workflow-multibranch</artifactId>
+      <version>1.14</version>
+      <scope>test</scope>
+    </dependency> 
+    <dependency>
```
Comment (benjaminfuchs) [inline pom.xml]: Needed for the updated/new tests.

File: src/test/java/hudson/plugins/warnings/WarningsPublisherTest.java [hunk 3]
```diff
@@ -20,32 +26,32 @@
  */
 public class WarningsPublisherTest {
     private static final String SUFFIX_NAME = " Warnings";
-    private static final String SECOND = "JSLint";
+    private static final String SECOND = "PyLint";
```
Comment (benjaminfuchs) [inline src/test/java/hudson/plugins/warnings/WarningsPublisherTest.java]: JSLint led to problem with console tests, PyLint worked fine.

**Comment by uhafner** [comment]
Thanks!

It would be helpful for this PR (and the corresponding analysis core PR) if you can also provide a small test case in the acceptance test harness.  

Did you try to run the acceptance test harness with the new snapshot releases?

**Comment by benjaminfuchs** [comment]
@uhafner 
I added the WarningsMultiBranchTest that check the Issue31202, i.e. that the graphs are shown. 
Is that sufficient or what else did you mean with "acceptance test harness", maybe I just didnt get it right.

I ran all snapshots together (by local maven releaes, analysis-pom, analysis-core, warnings-plugin) and it worked fine.

**Comment by uhafner** [comment]
https://github.com/jenkinsci/acceptance-test-harness
I think it makes sense to check that all tests are still running before a release. (It would be even better to have a new test case:-)

You can run the tests using the command `BROWSER=chrome mvn -Dtest=AnalysisSuite`.

**Comment by benjaminfuchs** [comment]
Okay I started the test harness, if my VM finishes the job I will let you now.

**Comment by benjaminfuchs** [comment]
@uhafner 
Had some busy time, but guess its time to get this moving forward.
As written in the analysis-core-plugin thread, Im stuck right now with some failures in the WarningsPluginTest of the acceptance-test-harness:
[plugins.WarningsPluginTest.txt](https://github.com/jenkinsci/warnings-plugin/files/509346/plugins.WarningsPluginTest.txt)
Will try to find some time the next days to dig in there.
Greetings

**Comment by uhafner** [comment]
Hmm, I also get test failures when I apply the changes. I'll need to investigate this in more detail...

**Comment by benjaminfuchs** [comment]
Hi, do you get the same failures I had? If you get the same errors I can be sure I’m not debugging false fails and could help out.

**Comment by uhafner** [comment]
I get:
```
Failed tests: 
plugins.WarningsPluginTest.should_navigate_to_result_action_from_job(plugins.WarningsPluginTest)
  Run 1: WarningsPluginTest>AbstractAnalysisTest.should_navigate_to_result_action_from_job:144 
Expected: Matches regexp warnings.*Result
     but: was "http://127.0.0.1:61030/job/technical_excess/warnings/"
  Run 2: WarningsPluginTest>AbstractAnalysisTest.should_navigate_to_result_action_from_job:144 
Expected: Matches regexp warnings.*Result
     but: was "http://127.0.0.1:57276/job/mutual_bat/warnings/"

plugins.WarningsPluginTest.should_set_warnings_count_in_list_view_column_for_matrix_project(plugins.WarningsPluginTest)
  Run 1: WarningsPluginTest.should_set_warnings_count_in_list_view_column_for_matrix_project:473->assertValidLink:485 
Expected: is <1>
     but: was <0>
  Run 2: WarningsPluginTest.should_set_warnings_count_in_list_view_column_for_matrix_project:473->assertValidLink:485 
Expected: is <1>
     but: was <0>


Tests in error: 
plugins.WarningsPluginTest.should_not_skip_failed_builds_with_option_run_always(plugins.WarningsPluginTest)
  Run 1: WarningsPluginTest.should_not_skip_failed_builds_with_option_run_always » Provision
  Run 2: WarningsPluginTest.should_not_skip_failed_builds_with_option_run_always:623->assertThatActionExists:684 
Expected: contains action titled Java Warnings
     but: <http://127.0.0.1:57037/job/dull_sink/> does not have action: "Java Warnings"

plugins.WarningsPluginTest.should_report_warnings_per_axis(plugins.WarningsPluginTest)
  Run 1: WarningsPluginTest.should_report_warnings_per_axis:320->assertThatConfigurationTabIsCorrectlyFilled:331->CapybaraPortingLayerImpl.find:209 » NoSuchElement
  Run 2: WarningsPluginTest.should_report_warnings_per_axis:320->assertThatConfigurationTabIsCorrectlyFilled:331->CapybaraPortingLayerImpl.find:209 » NoSuchElement

plugins.WarningsPluginTest.should_set_warnings_count_in_list_view_column_for_freestyle_project(plugins.WarningsPluginTest)
  Run 1: WarningsPluginTest.should_set_warnings_count_in_list_view_column_for_freestyle_project » Provision
  Run 2: WarningsPluginTest.should_set_warnings_count_in_list_view_column_for_freestyle_project:457->assertValidLink:485 
Expected: is <1>
     but: was <0>


Flaked tests: 
plugins.WarningsPluginTest.should_detect_warnings_of_multiple_compilers_in_console_freestyle(plugins.WarningsPluginTest)
  Run 1: WarningsPluginTest.should_detect_warnings_of_multiple_compilers_in_console_freestyle » Provision
  Run 2: PASS

plugins.WarningsPluginTest.should_not_detect_errors_in_ignored_parts_of_the_workspace(plugins.WarningsPluginTest)
  Run 1: WarningsPluginTest.should_not_detect_errors_in_ignored_parts_of_the_workspace » Provision
  Run 2: PASS

plugins.WarningsPluginTest.should_skip_failed_builds(plugins.WarningsPluginTest)
  Run 1: WarningsPluginTest.should_skip_failed_builds » Provision Guice provision error...
  Run 2: PASS


Tests run: 23, Failures: 2, Errors: 3, Skipped: 0, Flakes: 3

```

**Comment by uhafner** [comment]
I fixed the AggregatedAction, now only one failure is left (should_not_skip_failed_builds_with_option_run_always).

**Comment by benjaminfuchs** [comment]
Nice to here, I was playing around in interactive mode with that fail and saw, that if you have some passes after the fail the expected "Warnings" row appears. Maybe an interesting information.

**Comment by benjaminfuchs** [comment]
Nice to here, I was playing around in interactive mode with that fail and saw, that if you have some passes after the fail the expected "Warnings" row appears. Maybe an interesting information. 
Github seems not to be able to upload pictures in the moment ....

**Comment by benjaminfuchs** [comment]
ah, sorry wrong button...

**Comment by benjaminfuchs** [comment]
@uhafner At the moment I guess our problem is, that SimpleBuildStep.LastBuildAction is only called for successful builds:

```
List<Action> actions = new LinkedList<>();
Run r = j.getLastSuccessfulBuild();
if (r != null) {
    for (LastBuildAction a : r.getActions(LastBuildAction.class)) {
        actions.addAll(a.getProjectActions());
    }
}
// TODO should there be an option to check lastCompletedBuild even if it failed?
// Not useful for, say, TestResultAction, since if you have a build that fails before recording test
// results, the job would then have no TestResultProjectAction.
return actions;
```

https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/tasks/SimpleBuildStep.java

**Comment by uhafner** [comment]
Yes, I've also debugged into the same line :-( I sent a mail to the dev list, let's see what the core devs can do...

**Comment by uhafner** [comment]
There is already an issue: https://issues.jenkins-ci.org/browse/JENKINS-28479

**Comment by uhafner** [comment]
Ok, I merged your changes now. Hopefully, someone will fix https://issues.jenkins-ci.org/browse/JENKINS-28479 soon.

**Pull request closed without merging**